DataFormats/Detectors/CTP/src/ctpRateFetcher.cxx

```diff
@@ -0,0 +1,182 @@
+{
+  std::vector<ctp::CTPClass> ctpcls = mConfig->getCTPClasses();
+  std::vector<int> clslist = mConfig->getTriggerClassList();
+  int classIndex = -1;
+  for (size_t i = 0; i < clslist.size(); i++) {
+    if (ctpcls[i].name.find(className) != std::string::npos) {
+      classIndex = i;
+      break;
+    }
+  }
+  if (classIndex == -1) {
+    LOG(warn) << "Trigger class " << className << " not found in CTPConfiguration";
+    return -1.;
+  }
+  auto rate{mScalers->getRateGivenT(timeStamp * 1.e-3, classIndex, inputType)};
+  return rate.second;
+}
+double ctpRateFetcher::fetchCTPratesInputs(o2::ccdb::BasicCCDBManager* /*ccdb*/, uint64_t timeStamp, int /*runNumber*/, int input)
+{
+  std::vector<ctp::CTPScalerRecordO2> recs = mScalers->getScalerRecordO2();
+  if (recs[0].scalersInps.size() == 48) {
+    return pileUpCorrection(mScalers->getRateGivenT(timeStamp * 1.e-3, input, 7).second);
+  } else {
+    LOG(error) << "Inputs not available";
+    return -1.;
+  }
+}
+double ctpRateFetcher::fetchCTPratesInputsNoPuCorr(o2::ccdb::BasicCCDBManager* /*ccdb*/, uint64_t timeStamp, int /*runNumber*/, int input)
+{
+  std::vector<ctp::CTPScalerRecordO2> recs = mScalers->getScalerRecordO2();
+  if (recs[0].scalersInps.size() == 48) {
+    return mScalers->getRateGivenT(timeStamp * 1.e-3, input, 7).second;
+  } else {
+    LOG(error) << "Inputs not available";
+    return -1.;
+  }
+}
+
+double ctpRateFetcher::pileUpCorrection(double triggerRate)
+{
+  if (mLHCIFdata == nullptr) {
+    LOG(fatal) << "No filling" << std::endl;
+  }
+  auto bfilling = mLHCIFdata->getBunchFilling();
+  std::vector<int> bcs = bfilling.getFilledBCs();
+  double nbc = bcs.size();
+  double nTriggersPerFilledBC = triggerRate / nbc / constants::lhc::LHCRevFreq;
+  double mu = -std::log(1 - nTriggersPerFilledBC);
+  return mu * nbc * constants::lhc::LHCRevFreq;
+}
+
+void ctpRateFetcher::setupRun(int runNumber, o2::ccdb::BasicCCDBManager* ccdb, uint64_t timeStamp)
+{
+  if (runNumber == mRunNumber) {
+    return;
+  }
+  mRunNumber = runNumber;
+  LOG(debug) << "Setting up CTP scalers for run " << mRunNumber;
+  if (mManualCleanup) {
+    delete mConfig;
+    delete mScalers;
+    delete mLHCIFdata;
+  }
+  std::map<string, string> metadata;
+  mLHCIFdata = ccdb->getSpecific<parameters::GRPLHCIFData>("GLO/Config/GRPLHCIF", timeStamp, metadata);
+  if (mLHCIFdata == nullptr) {
+    LOG(fatal) << "GRPLHCIFData not in database, timestamp:" << timeStamp;
+  }
+  metadata["runNumber"] = std::to_string(mRunNumber);
+  mConfig = ccdb->getSpecific<ctp::CTPConfiguration>("CTP/Config/Config", timeStamp, metadata);
+  if (mConfig == nullptr) {
+    LOG(fatal) << "CTPRunConfig not in database, timestamp:" << timeStamp;
+  }
+  mScalers = ccdb->getSpecific<ctp::CTPRunScalers>("CTP/Calib/Scalers", timeStamp, metadata);
+  if (mScalers == nullptr) {
+    LOG(fatal) << "CTPRunScalers not in database, timestamp:" << timeStamp;
+  }
+  mScalers->convertRawToO2();
```

Comment:
Hi, 
Thank you, it goes in the right direction.

I understand that I have to `setupRun(runNumber, ccdbManager, timestamp, false)` when initializing. 
1. ccdbManager: Can I just pass the `CcdbManager`'s singleton ? I am still not sure why I have to pass it in such case, but I don't mind of course. 
2. timestamp: can I just pass the current timestamp, will that work ?

then, when I want to get a value out of some scalers I should do : 
```
updateScalers(ctp::CTPRunScalers* scalers) // if new scalers
auto valueForTheUsers = ctpRateFetcher::fetchNoPuCorr(ccdbManager, timestamp, runNumber, sourceName)
```
1. ccdbManager: same question as above
2. timestamp: current timestamp
3. Here, the issue is that I see that `fetchNoPuCorr` calls `setupRun` which will not work then because I must call it myself. Or am I mistaken ?

Reply:
It was done as suggested. Macro TestFetcher.C can help to clarify